### PR TITLE
34 soft/hard/normal negative preconditions policy support for sam and masam

### PR DESCRIPTION
**Pull request by oriKay**

### CODE REVIEW

File: sam_learning/learners/sam_learning.py [hunk 1]
```diff
@@ -218,6 +223,44 @@ def deduce_initial_inequality_preconditions(self) -> None:
...
+                    copy_precondition_positive.is_positive = True
+                    if copy_precondition_positive.untyped_representation in action_add_effects:
+                        continue
+
+                new_preconditions.add(precondition)
+
+            action.preconditions.root.operands = new_preconditions
+
+    def handle_negative_preconditions_policy(self):
+        match self.negative_preconditions_policy:
```
Comment (argaman-aloni) [inline sam_learning/learners/sam_learning.py:257]: Notice that switch case was only added in python 3.10 (as far as I know) and our code supports python 3.9 (at lease that's how I run the code)

File: sam_learning/learners/sam_learning.py [hunk 2]
```diff
@@ -218,6 +223,44 @@ def deduce_initial_inequality_preconditions(self) -> None:
                 if action_data.signature[lifted_param1] == action_data.signature[lifted_param2]:
                     action_data.preconditions.root.inequality_preconditions.add((lifted_param1, lifted_param2))
 
+    def hard_remove_negative_preconditions(self):
```
Comment (argaman-aloni) [inline sam_learning/learners/sam_learning.py:226]: merge the two functions, there is no need for them to exist independently, just check the policy before deleting....

File: sam_learning/learners/sam_learning.py [hunk 3]
```diff
@@ -218,6 +223,44 @@ def deduce_initial_inequality_preconditions(self) -> None:
...
+
+    def soft_remove_negative_preconditions(self):
+        """Removes negative preconditions when the fluent is an add effect"""
+        for action in self.partial_domain.actions.values():
+            new_preconditions = set()
+            action_add_effects = [effect.untyped_representation for effect in action.discrete_effects if effect.is_positive]
+
+            for precondition in action.preconditions.root.operands:
+                if isinstance(precondition, Predicate) and not precondition.is_positive:
+                    copy_precondition_positive = precondition.copy()
```
Comment (argaman-aloni) [inline sam_learning/learners/sam_learning.py:247]: I added to pddl-plus-parser==3.6.3 the option to call copy as follows: 
result = predicate.copy(is_negated=True)
and then you can remove the extra line where you set the is_positive value.

File: tests/multi_agent_learning_tests/multi_agent_sam_test.py [hunk 4]
```diff
@@ -251,3 +264,91 @@ def test_learn_action_model_returns_learned_model(
...
+
+        difference_ignore_from_classic = preconds_ignore.difference(preconds)
+        difference_classic_from_ignore = preconds.difference(preconds_ignore)
+        add_effects = [effect.untyped_representation for effect in action.discrete_effects if effect.is_positive]
+
+        assert len(difference_ignore_from_classic) == 0
+
+        for pre in difference_classic_from_ignore:
+            if isinstance(pre, Predicate):
+                pre_positive_copy = pre.copy()
```
Comment (argaman-aloni) [inline tests/multi_agent_learning_tests/multi_agent_sam_test.py:338]: use the same here (I also intend to use it to prettify my code a bit)

File: tests/multi_agent_learning_tests/multi_agent_sam_test.py [hunk 5]
```diff
@@ -251,3 +264,91 @@ def test_learn_action_model_returns_learned_model(
...
+                pre_positive_copy.is_positive = True
+                assert (not pre.is_positive) and (pre_positive_copy.untyped_representation in add_effects)
+
+
+def test_learn_action_model_hard_policy_returns_learned_model_no_negative_precondition(
+        woodworking_ma_sam_hard_policy: MultiAgentSAM, multi_agent_observation: MultiAgentObservation):
+    learned_model, learning_report = woodworking_ma_sam_hard_policy.learn_combined_action_model([multi_agent_observation])
+    print(learning_report)
+    print(learned_model.to_pddl())
```
Comment (argaman-aloni) [inline tests/multi_agent_learning_tests/multi_agent_sam_test.py:347]: You don't validate what you said you want to validate.... This is a good test to see that the learning process does not fail in general (I admit I have some of those) but if you say that you are validating something add the appropriate asserts.

File: tests/multi_agent_learning_tests/multi_agent_sam_test.py [hunk 6]
```diff
@@ -251,3 +264,91 @@ def test_learn_action_model_returns_learned_model(
...
+
+
+def test_learn_action_model_soft_policy_returns_learned_model_no_negative_precondition_has_add_effect(
+        woodworking_ma_sam_soft_policy: MultiAgentSAM, multi_agent_observation: MultiAgentObservation):
+    learned_model, learning_report = woodworking_ma_sam_soft_policy.learn_combined_action_model([multi_agent_observation])
+    print(learning_report)
```
Comment (argaman-aloni) [inline tests/multi_agent_learning_tests/multi_agent_sam_test.py:353]: same here.

File: tests/sam_learning_test.py [hunk 7]
```diff
@@ -426,3 +426,77 @@ def test_learn_action_model_returns_learned_model(elevators_sam_learning: SAMLea
     learned_model, learning_report = elevators_sam_learning.learn_action_model([elevators_observation])
     print(learning_report)
     print(learned_model.to_pddl())
+
+
+def test_learn_action_model_with_hard_policy_deletes_negative_preconditions(
```
Comment (argaman-aloni) [inline tests/sam_learning_test.py:431]: since the tests are the same as in MA-SAM apply the changes from there to here as well.

**Review by argaman-aloni — CHANGES_REQUESTED** [review]

File: sam_learning/learners/sam_learning.py (hunk 1 above)
Comment (oriKay) [reply sam_learning/learners/sam_learning.py:257]: I checked the readme doc first before using, and it says python 3.10+

Comment (argaman-aloni) [reply sam_learning/learners/sam_learning.py]: Really?  But my code is python 3.9.12, so it should work with my environment as well, I'll fix the readme.

File: sam_learning/learners/sam_learning.py [hunk 8]
```diff
@@ -218,6 +223,29 @@ def deduce_initial_inequality_preconditions(self) -> None:
...
+                    copy_precondition_positive = precondition.copy(is_negated=True)
+                    if ((not self.negative_preconditions_policy == NegativePreconditionPolicy.soft) or
+                            copy_precondition_positive.untyped_representation in action_add_effects):
+                        continue
+
+                new_preconditions.add(precondition)
+
+            action.preconditions.root.operands = new_preconditions
+
+    def handle_negative_preconditions_policy(self):
```
Comment (argaman-aloni) [inline sam_learning/learners/sam_learning.py:244]: just inline to the execution. it's a redundant function

File: utilities/util_types.py [hunk 9]
```diff
@@ -41,3 +41,9 @@ class SolutionOutputTypes(Enum):
     not_applicable = 4
     goal_not_achieved = 5
     solver_error = 6
+
+
+class NegativePreconditionPolicy(Enum):
+    normal = 1,
```
Comment (argaman-aloni) [inline utilities/util_types.py:47]: maybe instead of normal call this no_remove

**Review by argaman-aloni — CHANGES_REQUESTED** [review]

**Review by argaman-aloni — APPROVED** [review]

**Pull request merged**